### PR TITLE
Force and soft updates integrations

### PR DESCRIPTION
**Pull request by MahmoudHafezDarwish**

### CODE REVIEW

File: lib/core/remote_config/models/current_version.dart [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            ? null
+            : PlatformVersion.fromJson(json['android'] as Map<String, dynamic>),
+      );
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'ios': ios?.toJson(),
+        'android': android?.toJson(),
+      };
+
+  bool isSoftUpdate() {
```
Comment (salahamassi) [inline lib/core/remote_config/models/current_version.dart:30]: bool get isSoftUpdate =>

File: lib/core/remote_config/models/current_version.dart [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    var currentV = int.tryParse(sl<PackageInfo>().buildNumber);
+    {
+      if (Platform.isAndroid) {
+        if (currentV! < android!.maxVersion!) {
```
Comment (salahamassi) [inline lib/core/remote_config/models/current_version.dart:34]: decide android and ios props if they are nullable or not

File: lib/core/remote_config/models/current_version.dart [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          return true;
+        } else {
+          return false;
+        }
+      }
+    }
+  }
+
+  bool isForceUpdate() {
+    var currentV = int.tryParse(sl<PackageInfo>().buildNumber);
```
Comment (salahamassi) [inline lib/core/remote_config/models/current_version.dart:50]: why var?

Comment (salahamassi) [reply lib/core/remote_config/models/current_version.dart:50]: currentVersion**

File: lib/core/remote_config/models/current_version.dart [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    } else {
+      if (currentV! < ios!.minVersion!) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  String? getMessage() {
```
Comment (salahamassi) [inline lib/core/remote_config/models/current_version.dart:67]: String get message => Platform.isAndroid ? ...

File: lib/core/remote_config/models/platform_version.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class PlatformVersion {
+  const PlatformVersion({this.minVersion, this.maxVersion, this.message});
+
+  final int? minVersion;
```
Comment (salahamassi) [inline lib/core/remote_config/models/platform_version.dart:4]: why props are nullable?

File: lib/core/remote_config/page/soft_update_page.dart [hunk 6]
```diff
@@ -50,7 +46,7 @@ class _SoftUpdatePageState extends State<SoftUpdatePage> {
               Padding(
                 padding: const EdgeInsets.symmetric(horizontal: 87.0),
                 child: Text(
-                  widget.message,
+                  message,
```
Comment (salahamassi) [inline lib/core/remote_config/page/soft_update_page.dart:49]: context.localizations.update_app_text

Comment (salahamassi) [reply lib/core/remote_config/page/soft_update_page.dart:49]: decide message read from constructor or from localizations directly

File: lib/core/remote_config/remote_config_service.dart [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  String get buildNumber => sl<PackageInfo>().buildNumber;
+
+  void initRemoteConfig() async {
+    await FirebaseRemoteConfig.instance.setConfigSettings(RemoteConfigSettings(
+      fetchTimeout: const Duration(minutes: 1),
+      minimumFetchInterval: const Duration(minutes: 5),
+    ));
+
+    final defaults = <String, dynamic>{
+      'maintenance_mode': '0',
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:26]: remove first three keys

File: lib/core/remote_config/remote_config_service.dart [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    try {
+      debugPrint('Config :: start');
+      await FirebaseRemoteConfig.instance.tryfetchAndActivate();
+
+      await FirebaseRemoteConfig.instance.fetch();
+      debugPrint('Config :: start 2');
+
+      await FirebaseRemoteConfig.instance.activate();
+      var jsonStr =
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:44]: why var?

File: lib/core/remote_config/remote_config_service.dart [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+          FirebaseRemoteConfig.instance.getString('appCurrentVersion');
+      debugPrint('Config :: $jsonStr');
+      remoteConfigVer = CurrentVersion.fromJson(json.decode(jsonStr));
+      debugPrint('Config ::  Current st $remoteConfigVer}');
+
+      var currentV = int.tryParse(buildNumber);
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:50]: improve name and why var?

File: lib/core/remote_config/remote_config_service.dart [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      debugPrint('Config ::  Current $currentV');
+      if (remoteConfigVer!.isForceUpdate()) {
+        sl<AppRouter>()
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:53]: **sl<AppRouter>()** replace it with **appRouter**

File: lib/core/remote_config/remote_config_service.dart [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import 'package:firebase_remote_config/firebase_remote_config.dart';
+import 'package:flutter/material.dart';
+import 'package:one_studio_core/core.dart';
+import 'package:package_info_plus/package_info_plus.dart';
+
+import 'models/current_version.dart';
+
+class RemoteConfigService {
+  FirebaseRemoteConfig remoteConfig = FirebaseRemoteConfig.instance;
+  CurrentVersion? remoteConfigVer;
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:15]: remove this global variable

File: lib/core/remote_config/remote_config_service.dart [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      }
+    } catch (e) {
+      debugPrint('error $e');
+    }
+  }
+
+  var showSoftUpdate = true;
+
+  void checkAndShowSoftUpdate(BuildContext context) {
+    if (showSoftUpdate == false) {
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:64]: get FirebaseRemoteConfig.instance.getString('appCurrentVersion'); again

File: lib/core/remote_config/remote_config_service.dart [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import 'package:bond/core/helepers/extensions/auto_route_extensions.dart';
+import 'package:bond/routes/app_router.dart';
+import 'package:firebase_remote_config/firebase_remote_config.dart';
+import 'package:flutter/material.dart';
+import 'package:one_studio_core/core.dart';
+import 'package:package_info_plus/package_info_plus.dart';
+
+import 'models/current_version.dart';
+
+class RemoteConfigService {
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:13]: consturcter call init method

File: lib/core/remote_config/remote_config_service.dart [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  FirebaseRemoteConfig remoteConfig = FirebaseRemoteConfig.instance;
+  CurrentVersion? remoteConfigVer;
+
+  String get buildNumber => sl<PackageInfo>().buildNumber;
+
+  void initRemoteConfig() async {
```
Comment (salahamassi) [inline lib/core/remote_config/remote_config_service.dart:19]: rename it to call

File: lib/providers/app_service_provider.dart [hunk 15]
```diff
@@ -19,5 +20,7 @@ class AppServiceProvider extends ServiceProvider {
     );
 
     it.registerFactory(() => AppBloc(it()));
+    it.registerLazySingleton(() => RemoteConfigService());
+    sl<RemoteConfigService>().initRemoteConfig();
```
Comment (salahamassi) [inline lib/providers/app_service_provider.dart:24]: remove this line

File: lib/providers/app_service_provider.dart [hunk 16]
```diff
@@ -19,5 +20,7 @@ class AppServiceProvider extends ServiceProvider {
     );
 
     it.registerFactory(() => AppBloc(it()));
+    it.registerLazySingleton(() => RemoteConfigService());
```
Comment (salahamassi) [inline lib/providers/app_service_provider.dart:23]: singleton not LazySingleton

File: lib/core/remote_config/models/current_version.dart (hunk 1 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/models/current_version.dart:30]: done

File: lib/core/remote_config/models/current_version.dart (hunk 3 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/models/current_version.dart:50]: I change it for the final int? currentVersion

File: lib/core/remote_config/models/current_version.dart (hunk 4 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/models/current_version.dart:67]: done

File: lib/core/remote_config/models/platform_version.dart [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+  final int? maxVersion;
+  final String? message;
+
+  factory PlatformVersion.fromJson(Map<String, dynamic> json) =>
+      PlatformVersion(
+        minVersion: json['min_version'] as int? ?? 81,
```
[the remark this answers is not in the document]
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/models/platform_version.dart:10]: Ok i remove it

File: lib/core/remote_config/page/soft_update_page.dart (hunk 6 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/page/soft_update_page.dart:49]: final String? message;
I make it optional and I put the default value from localizations directly
                {  message ?? context.localizations.update_app_text, }

File: lib/core/remote_config/remote_config_service.dart (hunk 7 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:26]: I remove it

File: lib/core/remote_config/remote_config_service.dart (hunk 8 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:44]: مش صح يكون var حولته  ل String  و Final

File: lib/core/remote_config/remote_config_service.dart (hunk 9 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:50]: done I replace var with int?  and I improve name of variable

File: lib/core/remote_config/remote_config_service.dart (hunk 10 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:53]: Do you mean replacing it with context.router

File: lib/core/remote_config/remote_config_service.dart (hunk 11 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:15]: done

File: lib/core/remote_config/remote_config_service.dart (hunk 12 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:64]: done

File: lib/core/remote_config/remote_config_service.dart (hunk 14 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/remote_config_service.dart:19]: done

File: lib/providers/app_service_provider.dart (hunk 15 above)
Comment (MahmoudHafezDarwish) [reply lib/providers/app_service_provider.dart]: done

File: lib/core/remote_config/page/soft_update_page.dart [hunk 18]
```diff
@@ -7,18 +7,14 @@ import 'package:flutter/material.dart';
 import 'package:flutter_svg/flutter_svg.dart';
 import 'package:open_store/open_store.dart';
 
-class SoftUpdatePage extends StatefulWidget {
-  const SoftUpdatePage({Key? key, required this.message}) : super(key: key);
+class SoftUpdatePage extends StatelessWidget {
+  SoftUpdatePage({Key? key, required this.message}) : super(key: key);
 
-  final String message;
+  String message;
```
[the remark this answers is not in the document]
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/page/soft_update_page.dart]: done

File: lib/core/remote_config/models/platform_version.dart (hunk 5 above)
Comment (MahmoudHafezDarwish) [reply lib/core/remote_config/models/platform_version.dart:4]: can be null

File: lib/features/update_app_config/models/current_version.dart [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+  PlatformVersion get platformVersion {
+    if (Platform.isAndroid) {
+      return android;
+    } else {
+      return ios;
+    }
+  }
+
+  String? get message {
```
Comment (salahamassi) [inline lib/features/update_app_config/models/current_version.dart:34]: not nullable

File: lib/features/update_app_config/models/current_version.dart [hunk 20]
```diff
@@ -0,0 +1,41 @@
+import 'dart:io';
+
+import 'package:bond/features/update_app_config/models/platform_version.dart';
```
Comment (salahamassi) [inline lib/features/update_app_config/models/current_version.dart:3]: update_app_config -> update_app

File: lib/features/update_app_config/models/platform_version.dart [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+    required this.minVersion,
+    required this.maxVersion,
+    required this.message,
+  });
+
+  final int minVersion;
+  final int maxVersion;
+  final String message;
+
+  factory PlatformVersion.fromJson(Map<String, dynamic> json) =>
```
Comment (salahamassi) [inline lib/features/update_app_config/models/platform_version.dart:12]: use json serialize

File: lib/features/update_app_config/page/soft_update_page.dart [hunk 22]
```diff
@@ -7,16 +7,11 @@ import 'package:flutter/material.dart';
 import 'package:flutter_svg/flutter_svg.dart';
 import 'package:open_store/open_store.dart';
 
-class SoftUpdatePage extends StatefulWidget {
-  const SoftUpdatePage({Key? key, required this.message}) : super(key: key);
+class SoftUpdatePage extends StatelessWidget {
+  const SoftUpdatePage({Key? key, this.message}) : super(key: key);
 
-  final String message;
+  final String? message;
```
Comment (salahamassi) [inline lib/features/update_app_config/page/soft_update_page.dart:13]: not nullable

**Review by salahamassi — APPROVED** [review]

File: lib/features/update_app_config/models/current_version.dart (hunk 20 above)
Comment (MahmoudHafezDarwish) [reply lib/features/update_app_config/models/current_version.dart]: done

File: lib/features/update_app_config/models/current_version.dart (hunk 19 above)
Comment (MahmoudHafezDarwish) [reply lib/features/update_app_config/models/current_version.dart]: done

File: lib/features/update_app_config/page/soft_update_page.dart (hunk 22 above)
Comment (MahmoudHafezDarwish) [reply lib/features/update_app_config/page/soft_update_page.dart]: ok

File: lib/features/update_app_config/models/platform_version.dart (hunk 21 above)
Comment (MahmoudHafezDarwish) [reply lib/features/update_app_config/models/platform_version.dart]: done

**Pull request merged**